[FX lowering] Modify replace_all_uses_with to allowing filtering of nodes to update; use it to

jaybean-dev opened the pull request:
Summary:
The test that is enabled generates a graph as such:

```
linear_25 --> sigmoid_14 --> output_1
         \--> output_2
```
Before this diff, (unpadding) layout_transform nodes would be added as follows:

```
linear_25 --> layout_xform1 --> sigmoid_14 --> layout_xform2--> output_1
                           \--> output_2
```
This causes an assertion to fail for the sigmoid node where the input and output types
don't match due to padding differences.

This diff modifies the replacement algorithm to not affect users of an output's parent node
when the user requires padded inputs. This yields the following graph instead:

```
linear_25 --> sigmoid_14 --> layout_xform2--> output_1
         \--> layout_xform1 --> output_2
```

Test Plan: Manually and CI

Differential Revision: D34623590

jamesr66a reviewed (approved):
Looks good to me. I added a comment about rewording the name of the argument and the comment to make its function more clear


BTW you will need to run the backward compatibility test with `--accept` to squelch the error. The way you've added this parameter, kwarg with default, is sufficient for preserving BC, so you have my approval to `--accept` the change.

Thanks!

jaybean-dev commented on torch/fx/node.py:
```diff
@@ -461,20 +461,26 @@ def format_node(self,
         Args:
 
             replace_with (Node): The node to replace all uses of ``self`` with.
+            allow_replacement_fn (Callable): Function that takes a node and returns a boolean indicating if this node should be replacement.
```
Thanks for the feedback. How do I run the backward compat test?

jaybean-dev committed:
[FX lowering] Modify replace_all_uses_with to allowing filtering of nodes to update; use it to (#73763)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/73763

The test that is enabled generates a graph as such:

```
linear_25 --> sigmoid_14 --> output_1
         \--> output_2
```
Before this diff, (unpadding) layout_transform nodes would be added as follows:

```
linear_25 --> layout_xform1 --> sigmoid_14 --> layout_xform2--> output_1
                           \--> output_2
```
This causes an assertion to fail for the sigmoid node where the input and output types
don't match due to padding differences.

This diff modifies the replacement algorithm to not affect users of an output's parent node
when the user requires padded inputs. This yields the following graph instead:

```
linear_25 --> sigmoid_14 --> layout_xform2--> output_1
         \--> layout_xform1 --> output_2
```

Test Plan: Manually and CI

Reviewed By: jfix71, dborkovic

Differential Revision: D34623590

fbshipit-source-id: eabb03017844be182dbe6abcf4fb91dc8d67cb88
